In [1]:
# Data Manipulation and numerical Calculations
import pandas as pd
import numpy as np
import random

# Data visualizations
import matplotlib.pyplot as plt
import seaborn as sns

#to ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
%cd "C:\Users\Admin\Documents\Nhập môn khoa học dữ liệu\BTL"

C:\Users\Admin\Documents\Nhập môn khoa học dữ liệu\BTL


In [3]:
data = pd.read_csv("./Summer-Olympic-medals-1976-to-2008.csv",encoding='latin-1')

In [4]:
data.head(10)

City    Year     Sport Discipline           Event  \
0  Montreal  1976.0  Aquatics     Diving  3m springboard   
1  Montreal  1976.0  Aquatics     Diving  3m springboard   
2  Montreal  1976.0  Aquatics     Diving  3m springboard   
3  Montreal  1976.0  Aquatics     Diving  3m springboard   
4  Montreal  1976.0  Aquatics     Diving    10m platform   
5  Montreal  1976.0  Aquatics     Diving    10m platform   
6       NaN  1976.0  Aquatics     Diving    10m platform   
7  Montreal  1976.0  Aquatics     Diving  3m springboard   
8  Montreal  1976.0  Aquatics     Diving    10m platform   
9  Montreal  1976.0  Aquatics     Diving    10m platform   

                     Athlete Gender Country_Code        Country Event_gender  \
0            KÖHLER, Christa  Women          GDR   East Germany            W   
1        KOSENKOV, Aleksandr    Men          URS   Soviet Union            M   
2       BOGGS, Philip George    Men          USA  United States            M   
3   CAGNOTTO, Giorgio Franco    Men          ITA          Italy            M   
4     WILSON, Deborah Keplar  Women          USA  United States            W   
5          LOUGANIS, Gregory    Men          USA  United States            M   
6     VAYTSEKHOVSKAYA, Elena  Women          URS   Soviet Union            W   
7  POTTER-MCINGVALE, Cynthia  Women          USA  United States            W   
8             DIBIASI, Klaus    Men          ITA          Italy            M   
9          ALEINIK, Vladimir    Men          URS   Soviet Union            M   

    Medal  
0  Silver  
1  Bronze  
2    Gold  
3  Silver  
4  Bronze  
5  Silver  
6    Gold  
7  Bronze  
8     NaN  
9  Bronze

City - Thành phố nơi tổ chức Olympic

Year - Năm tổ chức

Sport - Thể loại thể thao (eg. Aquatics: môn dưới nước, 
Atheletics: điền kinh, ...)

Discipline - Môn thi đấu ứng với thể loại (eg. Diving: nhảy cầu, Swimming: bơi etc.)

Event - Hạng mục thi đấu (eg. 3m springboard: nhảy cầu 3m, 10m platform etc.)

Athlete - Tên vận động viên giành huy chương

Gender - Giới tính

Country_Code - Mã quốc gia 3 ký tự

Country - Quốc tịch của vận động viên

Event_gender - Giới tính vận động viên khi tham gia tranh tài 
(Male, Female or Common Event)

Medal - Huy chương (eg. Bronze, Silver, Gold)

In [36]:
data.shape

(15317, 11)

- Dữ liệu
 + 15317 dòng
 + 11 cột 

# 1. Xử lí dữ liệu

1. Kiểm tra vấn đề dữ liệu
Một số vấn đề dữ liệu ta sẽ thực hiện kiểm tra:

Missing value

Duplicated

Lỗi logic

Bất đồng bộ

### 1.1 Kiểm tra dữ liệu bị thiếu

#### a. Kiểm tra trên bảng

In [37]:
data.isnull()

City   Year  Sport  Discipline  Event  Athlete  Gender  Country_Code  \
0      False  False  False       False  False    False   False         False   
1      False  False  False       False  False    False   False         False   
2      False  False  False       False  False    False   False         False   
3      False  False  False       False  False    False   False         False   
4      False  False  False       False  False    False   False         False   
...      ...    ...    ...         ...    ...      ...     ...           ...   
15312  False  False  False       False  False    False   False         False   
15313  False  False  False       False  False    False   False         False   
15314  False  False  False       False  False    False   False         False   
15315  False  False  False       False  False    False   False         False   
15316  False  False  False       False  False    False   False         False   

       Country  Event_gender  Medal  
0        False         False  False  
1        False         False  False  
2        False         False  False  
3        False         False  False  
4        False         False  False  
...        ...           ...    ...  
15312    False         False  False  
15313    False         False  False  
15314    False         False  False  
15315    False         False   True  
15316    False         False  False  

[15317 rows x 11 columns]

In [38]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15317 entries, 0 to 15316
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   City          15241 non-null  object 
 1   Year          15293 non-null  float64
 2   Sport         15317 non-null  object 
 3   Discipline    15317 non-null  object 
 4   Event         15316 non-null  object 
 5   Athlete       15317 non-null  object 
 6   Gender        15316 non-null  object 
 7   Country_Code  15317 non-null  object 
 8   Country       15238 non-null  object 
 9   Event_gender  15316 non-null  object 
 10  Medal         15271 non-null  object 
dtypes: float64(1), object(10)
memory usage: 1.3+ MB


**Một số thông tin thu được sau khi thực hiện kiếm tra:**
- Các cột City,Year,Country,Event_gender,Medal xuất  hiện giá trị bị thiếu
- Các cột có kiểu dữ liêu định lượng là:Year
- Các cột có kiểu dữ liệu định tính: City,Sport,Discipline,Event,Athlete,Gender,Country_Code,Country,Event_gender,Medal 

**Đếm số lượng missing value trong 1 cột**

In [39]:
data.isnull().sum()

City            76
Year            24
Sport            0
Discipline       0
Event            1
Athlete          0
Gender           1
Country_Code     0
Country         79
Event_gender     1
Medal           46
dtype: int64

In [40]:
(data.isnull().sum()/(len(data)))*100

City            0.496181
Year            0.156689
Sport           0.000000
Discipline      0.000000
Event           0.006529
Athlete         0.000000
Gender          0.006529
Country_Code    0.000000
Country         0.515767
Event_gender    0.006529
Medal           0.300320
dtype: float64

**Một số thông tin thu được sau khi thực hiện kiếm tra:**
- Cột City bị thiếu 76 dữ liệu

        => Giải pháp: Tính toán tỉ lệ giá trị bị thiếu trên tổng quan sát của mỗi cột. Có thể thấy dữ liệu tại cột City gặp vấn đề đến ~50% quan sát gặp vấn đề bị thiếu 
        => Thay thế bằng các gía trị xuất hiện nhiều nhất.
- Cột Year  bị thiếu 24 dữ liệu

        => Giải pháp: Bị mất gần 16% dữ liệu => Thay thế bằng các gía trị đứng trước nó.
- Cột Event bị thiếu 1 dữ liệu

        => Giải pháp: Bị mất gần 1% dữ liệu => Thay thế bằng các gía trị đứng trước nó.
- Cột Gender bị thiếu 1 dữ liệu

        => Giải pháp: Bị mất gần 1% dữ liệu => Thay thế bằng các gía trị đứng trước nó.
- Cột Country bị thiếu 79 dữ liệu

        => Giải pháp: Bị mất gần 52% dữ liệu => Thay thế bằng các gía trị xuất hiện nhiều nhất.
- Cột Medal bị thiếu 46 dữ liệu

        => Giải pháp: Bị mất 30% dữ liệu => Thay thế bằng các gía trị xuất hiện nhiều nhất.

### 1.2. Kiểm tra dữ liệu bị lặp

In [41]:
data.duplicated().sum()

3

- Có 3 cột bị lặp dữ liệu

In [42]:
data[data.duplicated()]

City    Year          Sport       Discipline  \
463   Montreal  1976.0  Canoe / Kayak  Canoe / Kayak F   
464   Montreal  1976.0  Canoe / Kayak  Canoe / Kayak F   
2205    Moscow  1980.0         Hockey           Hockey   

                        Event              Athlete Gender Country_Code  \
463   K-1 500m (kayak single)  KORSHUNOVA, Tatyana  Women          URS   
464   K-1 500m (kayak single)  KORSHUNOVA, Tatyana  Women          URS   
2205                   hockey         SINGH, Singh    Men          IND   

           Country Event_gender   Medal  
463   Soviet Union            W  Silver  
464   Soviet Union            W  Silver  
2205         India            M    Gold

In [43]:
data[data.duplicated()]

City    Year          Sport       Discipline  \
463   Montreal  1976.0  Canoe / Kayak  Canoe / Kayak F   
464   Montreal  1976.0  Canoe / Kayak  Canoe / Kayak F   
2205    Moscow  1980.0         Hockey           Hockey   

                        Event              Athlete Gender Country_Code  \
463   K-1 500m (kayak single)  KORSHUNOVA, Tatyana  Women          URS   
464   K-1 500m (kayak single)  KORSHUNOVA, Tatyana  Women          URS   
2205                   hockey         SINGH, Singh    Men          IND   

           Country Event_gender   Medal  
463   Soviet Union            W  Silver  
464   Soviet Union            W  Silver  
2205         India            M    Gold

- Dữ liệu bị lặp taị năm 1976 

  => Giaỉ pháp: Xóa 1 dòng
- Tại năm 1980 bị lặp môn thi "Hockey" taị cột Sport và Event  

  => Giaỉ pháp: Thay thế "Hockey" taị cột Sport và Event bằng giá trị đứng trước nó.

# 2.Xử lí dữ liệu 

## 2.1 Xử lí dữ liệu thiếu  

- Cột 'Year' đang ở dạng float nên ta sẽ ép kiểu về số nguyên 

In [44]:

data['Year'] = data['Year'].astype('int64')
data['Year'] 

IntCastingNaNError: Cannot convert non-finite values (NA or inf) to integer

In [ ]:
data['Year'].value_counts()

In [19]:
def update_city(data,nam=0,city=''):
    data[data['Year'] == nam].City.fillna(city, inplace=True)
# City là noiw tổ chức Olympic nên mỗi năm chỉ tổ chức ở 1 thành phố nên ta lọc được thành phố theo năm và 
#lấy mode() để lấy được tên nước vì ko biết lấy tên nước ra như nao

In [20]:
for i in range (1976,2009,4):
    city = data[data['Year'] == i].City.mode()[0]
    update_city(data,i,city)

In [22]:
data.isnull().sum()

City            76
Year            24
Sport            0
Discipline       0
Event            1
Athlete          0
Gender           1
Country_Code     0
Country         79
Event_gender     1
Medal           46
dtype: int64

In [50]:
data.City.isnull().sum()

76

#### Sửa cột Country 

In [5]:
x= data[data.Country.isnull()]['Country_Code'].unique()
x

array(['USA', 'GDR', 'FIN', 'POL', 'FRG', 'DEN', 'SUI', 'HUN', 'AUS',
       'URS', 'KOR', 'GRE', 'TCH', 'BRA', 'YUG', 'SWE', 'CHN', 'GER',
       'EUN', 'FRA', 'ROU', 'NOR', 'JPN', 'ESP', 'ITA', 'CZE', 'RUS',
       'NGR', 'RSA', 'LTU', 'CUB'], dtype=object)

In [21]:
def update_country(data,country_code=''):
    data[data['Country_Code'] == country_code].Country.fillna(data[data['Country_Code'] == country_code].Country.mode()[0],
                                                              inplace=True)

In [22]:
for i in x:
    a =data[data['Country_Code'] == i].Country
    a.fillna(a.mode()[0])

In [23]:
data[data['Country_Code'] == 'USA'].Country.isnull().sum()

11

In [13]:
data[data['Country_Code'] == 'USA'].Country.mode()[0]

'United States'

In [88]:
data.City.value_counts()

China          2107
Athens         1998
Sydney         1991
Atlanta        1850
Barcelona      1695
Seoul          1541
Los Angeles    1454
Moscow         1372
Montreal       1285
Russia           14
Canada           10
Name: City, dtype: int64

In [19]:
data['Year'].fillna(method = 'ffill')

0        1976.0
1        1976.0
2        1976.0
3        1976.0
4        1976.0
          ...  
15312    2008.0
15313    2008.0
15314    2008.0
15315    2008.0
15316    2008.0
Name: Year, Length: 15317, dtype: float64

In [20]:
data['Event'].fillna(method = 'ffill')

0        3m springboard
1        3m springboard
2        3m springboard
3        3m springboard
4          10m platform
              ...      
15312         66 - 74kg
15313         66 - 74kg
15314         55 - 60kg
15315         60 - 66kg
15316        96 - 120kg
Name: Event, Length: 15317, dtype: object

In [21]:
data['Gender'].fillna(method = 'ffill')

0        Women
1          Men
2          Men
3          Men
4        Women
         ...  
15312      Men
15313      Men
15314      Men
15315      Men
15316      Men
Name: Gender, Length: 15317, dtype: object

In [23]:
x = random.choice(data['Medal'].fillna(data['Medal'].mode()))
x

'Gold'

In [24]:
data.isnull().sum()

City            76
Year            24
Sport            0
Discipline       0
Event            1
Athlete          0
Gender           1
Country_Code     0
Country         79
Event_gender     1
Medal           46
dtype: int64

In [25]:
data['City'].fillna(data['City'].mode())

0        Montreal
1        Montreal
2        Montreal
3        Montreal
4        Montreal
           ...   
15312       China
15313       China
15314       China
15315       China
15316       China
Name: City, Length: 15317, dtype: object

In [26]:
data.value_counts("City")

City
China          2031
Athens         1998
Sydney         1991
Atlanta        1850
Barcelona      1695
Seoul          1541
Los Angeles    1454
Moscow         1372
Montreal       1285
Russia           14
Canada           10
dtype: int64

In [27]:
data['City'].fillna(method = 'ffill')

0        Montreal
1        Montreal
2        Montreal
3        Montreal
4        Montreal
           ...   
15312       China
15313       China
15314       China
15315       China
15316       China
Name: City, Length: 15317, dtype: object

In [28]:
data.isnull().sum()

City            76
Year            24
Sport            0
Discipline       0
Event            1
Athlete          0
Gender           1
Country_Code     0
Country         79
Event_gender     1
Medal           46
dtype: int64

## 2.2 Xử lí dữ liệu lặp 

In [90]:
data.drop_duplicates(keep='last',ignore_index=False,inplace = True)

In [93]:
data.duplicated().sum()

0